test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -258,29 +223,203 @@ public void multipleSubscriberCBRTest() throws Exception {
+        assertEquals(1, controller.listSubscribers(SCOPE_1, STREAM_2).join().size());
+
+        // Recreates the reader
+        reader = clientFactory.createReader(READER_GROUP_3 + "-" + 1, READER_GROUP_3, new JavaSerializer<>(),
+                ReaderConfig.builder().build(), clock::get, clock::get);
+        reader2 = clientFactory.createReader(READER_GROUP_4 + "-" + 1, READER_GROUP_4, new JavaSerializer<>(),
+                ReaderConfig.builder().build(), clock::get, clock::get);
+
+        // fill stream with 3 events
+        writingEventsToStream(3, writer, SCOPE_1, STREAM_1);
+        // fill stream with 4 events
+        writingEventsToStream(4, writer2, SCOPE_1, STREAM_2);
+
+        //Read 3 events with reader from the segment offset
+        readingEventsFromStream(3, reader);
+        readingEventsFromStream(3, reader2);
+
+        log.info("{} generating 2nd stream-cuts for {}/{}", READER_GROUP_3, SCOPE_1, STREAM_1);
+        streamCuts = generateStreamCuts(readerGroup, reader, clock);
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_3, streamCuts);
+        readerGroup.updateRetentionStreamCut(streamCuts);
+
+        log.info("{} generating 2nd stream-cuts for {}/{}", READER_GROUP_4, SCOPE_1, STREAM_2);
+        streamCuts2 = generateStreamCuts(readerGroup2, reader2, clock);
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_4, streamCuts2);
+        readerGroup2.updateRetentionStreamCut(streamCuts2);
+
+        // Retention set has two stream cut at 0/360, 0/450
+        // READER_GROUP_3 updated stream cut at 0/300, Subscriber lower bound is 0/300
+        // So truncating at stream cut 0/300 which leaves more data in the stream than min limit
+        // The timeout is set to 2 minutes a little longer than the retention period which is set to 1 minutes
+        // in order to confirm that the retention has taken place.
+        // Check to make sure truncation happened at SLB
+        AssertExtensions.assertEventuallyEquals("Truncation did not take place at offset 300.", true, () -> controller.getSegmentsAtTime(
+                        new StreamImpl(SCOPE_1, STREAM_1), 0L).join().values().stream().anyMatch(off -> off == 300),
+                5000, 2 * 60 * 1000L);
+        // Retention set has two stream cuts at 0/450...0/570
+        // READER_GROUP_4 updated stream cut at 0/390, Subscriber lower bound is 0/390
+        // Since 0/450 is the stream cut satisfying min and max bounds and SLB is lower than this.
+        // Truncation should happen at SLB
+        assertEquals(true, controller.getSegmentsAtTime(
+                new StreamImpl(SCOPE_1, STREAM_2), 0L).join().values().stream().anyMatch(off -> off == 390));
+    }
+
+
+    private void writingEventsToStream(int numberOfEvents, EventStreamWriter<String> writer, String scope, String stream) {
+        for (int i = 0; i < numberOfEvents; i++) {
```

Comment:
nit: `for (int i = 0; i < numberOfEvents; i++)` -> `for (int event = 0; event < numberOfEvents; event++)`

Reply:
done